File: app/src/main/java/at/shockbytes/dante/network/google/gson/GoogleBooksSuggestionResponseDeserializer.java
package at.shockbytes.dante.network.google.gson;

import com.google.gson.JsonArray;
import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParseException;

import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.List;

import at.shockbytes.dante.util.AppParams;
import at.shockbytes.dante.util.books.Book;
import at.shockbytes.dante.util.books.BookSuggestion;


/**
 * @author Martin Macheiner
 *         Date: 13.02.2017.
 */

public class GoogleBooksSuggestionResponseDeserializer implements JsonDeserializer<BookSuggestion> {

    @Override
    public BookSuggestion deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context)
            throws JsonParseException {

        if (json.getAsJsonObject().getAsJsonArray("items") != null) {

            int size = json.getAsJsonObject().getAsJsonArray("items").size();
            if (size > 0) {

                JsonObject volumeInfoMain = json.getAsJsonObject().getAsJsonArray("items").get(0)
                        .getAsJsonObject().get("volumeInfo").getAsJsonObject();

                Book mainSuggestion = grabBook(volumeInfoMain);
                List<Book> otherSuggestions = new ArrayList<>();

                size = (size >= AppParams.INSTANCE.getMaxFetchAmount())
                        ? AppParams.INSTANCE.getMaxFetchAmount()
                        : size;
                for (int i = 1; i < size; i++) {
                    JsonObject volumeInfo = json.getAsJsonObject().getAsJsonArray("items").get(i)
                            .getAsJsonObject().get("volumeInfo").getAsJsonObject();
                    otherSuggestions.add(grabBook(volumeInfo));
                }

                return new BookSuggestion(mainSuggestion, otherSuggestions);
            }
        }
        return null;
    }

    private Book grabBook(JsonObject volumeInfo) {

        String title = volumeInfo.get("title").getAsString();
        String subtitle = getSubtitle(volumeInfo);
        String author = getAuthors(volumeInfo);
        int pageCount = getPageCount(volumeInfo);
        String publishedDate = getPublishedDate(volumeInfo);
        String isbn = getIsbn(volumeInfo);
        String thumbnailAddress = getImageLink(volumeInfo);
        String googleBooksLink = getGoogleBooksLink(volumeInfo);
        String language = getLanguage(volumeInfo);

        return new Book(title, subtitle, author, pageCount, publishedDate,
                isbn, thumbnailAddress, googleBooksLink, language);
    }

    private String getPublishedDate(JsonObject volumeInfo) {
        if (volumeInfo.get("publishedDate") != null) {
            return volumeInfo.get("publishedDate").getAsString();
        }
        return "";
    }

    private String getGoogleBooksLink(JsonObject volumeInfo) {
        if (volumeInfo.get("infoLink") != null) {
            return volumeInfo.get("infoLink").getAsString();
        }
        return "";
    }

    private String getLanguage(JsonObject volumeInfo) {
        if (volumeInfo.get("language") != null) {
            return volumeInfo.get("language").getAsString();
        }
        return "";
    }

    private String getSubtitle(JsonObject volumeInfo) {

        if (volumeInfo.get("subtitle") != null) {
            return volumeInfo.get("subtitle").getAsString();
        }
        return "";
    }

    private int getPageCount(JsonObject volumeInfo) {
        if (volumeInfo.get("pageCount") != null) {
            return volumeInfo.get("pageCount").getAsInt();
        }
        return 0;
    }

    private String getIsbn(JsonObject volumeInfo) {

        if (volumeInfo.get("industryIdentifiers") != null) {
            JsonArray isbns = volumeInfo.get("industryIdentifiers").getAsJsonArray();
            for (JsonElement je : isbns) {
                JsonObject j = je.getAsJsonObject();
                if (j.get("type") != null && j.get("type").getAsString().equals("ISBN_13")) {
                    return j.get("identifier").getAsString();
                }
            }
        }
        return "";
    }

    private String getAuthors(JsonObject volumeInfo) {

        StringBuilder sb = new StringBuilder();
        if (volumeInfo.get("authors") != null) {
            JsonArray authors = volumeInfo.get("authors").getAsJsonArray();
            for (int i = 0; i < authors.size(); i++) {
                sb.append(authors.get(i).getAsString());
                if (i < authors.size() - 1) {
                    sb.append(", ");
                }
            }
        }
        return sb.toString();
    }

    private String getImageLink(JsonObject volumeInfo) {
        if (volumeInfo.get("imageLinks") != null) {
            return volumeInfo.get("imageLinks").getAsJsonObject()
                    .get("thumbnail").getAsString();
        }
        return null;
    }

}

File: app/src/main/java/at/shockbytes/dante/network/google/gson/GoogleBooksSuggestionResponseDeserializer.kt
package at.shockbytes.dante.network.google.gson

import at.shockbytes.dante.books.BookFactory
import at.shockbytes.dante.books.BookSuggestion
import at.shockbytes.dante.util.DanteUtils
import at.shockbytes.dante.util.books.Book
import com.google.gson.*
import java.lang.reflect.Type


/**
 * @author Martin Macheiner
 * Date: 13.02.2017.
 */

class GoogleBooksSuggestionResponseDeserializer : JsonDeserializer<BookSuggestion> {

    @Throws(JsonParseException::class)
    override fun deserialize(json: JsonElement, typeOfT: Type,
                             context: JsonDeserializationContext): BookSuggestion? {

        if (json.asJsonObject.getAsJsonArray("items") != null) {

            var size = json.asJsonObject.getAsJsonArray("items").size()
            if (size > 0) {

                // Retrieve the main information from the json object
                val volumeInfoMain = json.asJsonObject.getAsJsonArray("items").get(0)
                        .asJsonObject.get("volumeInfo").asJsonObject

                // Look for main suggestion and check for fetching size
                val mainSuggestion = grabBook(volumeInfoMain)
                size = if (size >= DanteUtils.maxFetchAmount) DanteUtils.maxFetchAmount else size

                // Because the data of other books is already fetched, let's convert them into objects
                val otherSuggestions = (1 until size).map { idx ->
                    val volumeInfo = json.asJsonObject.getAsJsonArray("items")
                            .get(idx).asJsonObject.get("volumeInfo").asJsonObject
                    grabBook(volumeInfo)
                }
                return BookSuggestion(mainSuggestion, otherSuggestions)
            }
        }
        return null
    }

    private fun grabBook(volumeInfo: JsonObject): Book {

        val title = volumeInfo.get("title").asString
        val subtitle = getSubtitle(volumeInfo)
        val author = getAuthors(volumeInfo)
        val pageCount = getPageCount(volumeInfo)
        val publishedDate = getPublishedDate(volumeInfo)
        val isbn = getIsbn(volumeInfo)
        val thumbnailAddress = getImageLink(volumeInfo)
        val googleBooksLink = getGoogleBooksLink(volumeInfo)
        val language = getLanguage(volumeInfo)

        return BookFactory.resolve(title, subtitle, author, pageCount, publishedDate, isbn,
                thumbnailAddress, googleBooksLink, language)
    }

    private fun getPublishedDate(volumeInfo: JsonObject): String {
        return if (volumeInfo.get("publishedDate") != null) {
            volumeInfo.get("publishedDate").asString
        } else ""
    }

    private fun getGoogleBooksLink(volumeInfo: JsonObject): String {
        return if (volumeInfo.get("infoLink") != null) {
            volumeInfo.get("infoLink").asString
        } else ""
    }

    private fun getLanguage(volumeInfo: JsonObject): String {
        return if (volumeInfo.get("language") != null) {
            volumeInfo.get("language").asString
        } else ""
    }

    private fun getSubtitle(volumeInfo: JsonObject): String {
        return if (volumeInfo.get("subtitle") != null) {
            volumeInfo.get("subtitle").asString
        } else ""
    }

    private fun getPageCount(volumeInfo: JsonObject): Int {
        return if (volumeInfo.get("pageCount") != null) {
            volumeInfo.get("pageCount").asInt
        } else 0
    }

    private fun getIsbn(volumeInfo: JsonObject): String {
        if (volumeInfo.get("industryIdentifiers") != null) {
            val allIsbn = volumeInfo.get("industryIdentifiers").asJsonArray
            allIsbn
                    .map { it.asJsonObject }
                    .filter { it.get("type") != null && it.get("type").asString == "ISBN_13" }
                    .forEach { return it.get("identifier").asString }
        }
        return ""
    }

    private fun getAuthors(volumeInfo: JsonObject): String {
        return if (volumeInfo.get("authors") != null) {
            val authors = volumeInfo.get("authors").asJsonArray
            authors.joinToString(", ") { it.asString }
        } else {
            ""
        }
    }

    private fun getImageLink(volumeInfo: JsonObject): String? {
        return if (volumeInfo.get("imageLinks") != null) {
            volumeInfo.get("imageLinks").asJsonObject
                    .get("thumbnail").asString
        } else null
    }

}
